Adding Maximum and Minimum reading_rate to readings tables

taylon4 opened the pull request:
# Description

This change adds a maximum and minimum reading rate for the daily/hourly readings unit table. This would ideally be used as a visual display on the front end when graphing data.

Fixes #594

## Type of change

(Check the ones that apply by placing an "x" instead of the space in the [ ] so it becomes [x])

- [ ] Note merging this changes the node modules
- [x] Note merging this changes the database configuration.
- [x] This change requires a documentation update

## Checklist

(Note what you have done by placing an "x" instead of the space in the [ ] so it becomes [x]. It is hoped you do all of them.)

- [x] I have followed the [OED pull request](https://openenergydashboard.github.io/developer/pr.html) ideas
- [ ] I have removed text in ( ) from the issue request

## Limitations

Edge cases should be tested with the max/min from the database. Specifically overlapping days - for example June 3rd 12:00 PM to June 4th 2:00 AM

taylon4 committed:
created test table to test the table

taylon4 committed:
Create test table

taylon4 committed:
Merge branch 'MaxMin' of https://github.com/taylon4/OED into MaxMin

taylon4 committed:
Added new table for max/min on daily_readings_unit

taylon4 committed:
add max/min column for hourly_readings_unit

taylon4 committed:
comments

taylon4 committed:
Add database changes to migrations

huss commented on src/server/migrations/0.8.0-1.0.0/sql/readings/create_reading_views.sql:
```diff
@@ -2,9 +2,88 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
+/*
```
This seems to be a copy of the SQL file. What it should have is any needed changes. So, date_trunc_up, etc. should not be present as those functions did not change. I think you only need to two views you changed.

huss replied:
I fixed this up but also noticed the migrations were off so I also edited so it has all the correct migrations for this file.

huss commented on src/server/sql/reading/create_reading_views.sql:
```diff
@@ -127,7 +127,76 @@ daily_readings_unit
-		END AS reading_rate,
+		END AS reading_rate, 
+		CASE WHEN u.unit_represent = 'quantity'::unit_represent_type THEN
+    	(max(( --Extract the maximum rate over each day
```
The indenting seems off on this line.

Also, it might be nice to have a comment that explains why the max includes the division (second extract) whereas the sum does not.

These apply to all cases following.

huss replied:
I addressed this. The part about second extract is unclear (even thought I made the comment) so ignored.

huss commented on src/server/sql/reading/create_reading_views.sql:
```diff
@@ -127,7 +127,76 @@ daily_readings_unit
+									-
+									greatest(r.start_timestamp, gen.interval_start)
+					)
+			)) )
```
I think the first ) should be on a separate line and lined up in the appropriate column to match the sum code above.

huss replied:
addressed,

huss commented on src/server/sql/reading/create_reading_views.sql:
```diff
@@ -127,7 +127,76 @@ daily_readings_unit
+									greatest(r.start_timestamp, gen.interval_start)
+					)
+			)) )
+			WHEN (u.unit_represent = 'flow'::unit_represent_type OR u.unit_represent = 'raw'::unit_represent_type) THEN
```
Indenting seems off. Maybe all can be checked.

huss commented on src/server/sql/reading/create_reading_views.sql:
```diff
@@ -311,7 +448,6 @@ BEGIN
 END;
 $$ LANGUAGE 'plpgsql';
 
```
I don't mind removing the double blank line here but then maybe do above to be consistent?

huss replied:
tried to make them all one blank line.

huss commented on src/server/sql/reading/create_reading_views.sql:
```diff
@@ -237,7 +374,7 @@ BEGIN
 	requested_range := shrink_tsrange_to_real_readings(tsrange(start_stamp, end_stamp, '[]'));
 	requested_interval := upper(requested_range) - lower(requested_range);
-	-- unit_column holds the column index into the cik table. This is the unit that was requested for graphing.
+	-- holds the column index into the cik table. This is the unit that was requested for graphing.
```
I'm not clear on why this comment was changed. Can you help me understand?

huss replied:
decided change was okay.

huss commented on src/server/migrations/0.8.0-1.0.0/sql/readings/create_reading_views.sql:
```diff
@@ -2,9 +2,88 @@
+This is used by the line graph function below to make them faster since the values
 are already averaged. There are two types of readings: quantity and flow/raw. The quantity
 readings must be normalized by their time length. The flow/raw readings are already by time
 so they are just averaged. The one table contains both types of readings but are now equivalent
```
While you did not do the first version, I think there should be
DROP MATERIALIZED VIEW IF EXISTS daily_readings_unit;
before the create so the create runs. The view normally should not be there but this would be safer. Same applies to hourly.

huss replied:
I realized you don't need to do this since this is a new view and the old one is dropped elsewhere.

huss reviewed (changes requested):
Thanks to @taylon4 for their first PR for OED. I have run some basic checks where the readings are all the same length and the results are as expected so the SQL seems correct. I have made a few comments that I would like addressed.

So everyone understands, I plan to comment out the new code and migration change when this PR is approved. It does not reflect in any way on this work. The issue is that OED cannot use the new max/min values until the graphics are changed and adding it to the view will slow down the DB during refreshes and take up space. The hope/plan is to get that done this summer and then the new code will be uncommented for inclusion. It is also hoped to have automated tests including readings that cross time boundaries by that time so all the SQL code can be tested (including this change) to verify it works as expected. Since this code uses the same logic as other places it is not expected to be an issue as manual tests in the past did not find an issue.

huss committed:
Merge remote-tracking branch 'upstream/development' into pr/taylon4/883

huss committed:
work on indenting

huss committed:
more formatting

huss committed:
migration & comment out

This comments out the min/max code as it is not yet used in the
front-end and it would slow it down/take space.
It also fixes (hopefully) the reading migration that was not complete
before this PR.
